crates/wasmtime/src/store/func_refs.rs

```diff
@@ -0,0 +1,76 @@
+use crate::module::ModuleRegistry;
+use std::{ptr::NonNull, sync::Arc};
+use wasmtime_runtime::{VMCallerCheckedFuncRef, VMNativeCallHostFuncContext};
+
+/// An arena of `VMCallerCheckedFuncRef`s.
+///
+/// Allows a store to pin and own funcrefs so that it can patch in trampolines
+/// for `VMCallerCheckedFuncRef`s that are missing a `wasm_call` trampoline and
+/// need Wasm to supply it.
+#[derive(Default)]
+pub struct FuncRefs {
+    /// A bump allocation arena where we allocate `VMCallerCheckedFuncRef`s such
+    /// that they are pinned and owned.
+    bump: bumpalo::Bump,
+
+    /// Pointers into `self.bump` for entries that need `wasm_call` field filled
+    /// in.
+    with_holes: Vec<NonNull<VMCallerCheckedFuncRef>>,
+
+    /// Pinned `VMCallerCheckedFuncRef`s that had their `wasm_call` field
+    /// pre-patched when constructing an `InstancePre`, and which we need to
+    /// keep alive for our owning store's lifetime.
+    instance_pre_func_refs: Vec<Arc<[VMCallerCheckedFuncRef]>>,
+}
+
+// Safety: this is safe because we only ever access `refs` when we have `&mut
+// self` and because the `VMCallerCheckedFuncRef`s we hand out are only ever
+// used on the same thread as `self`.
+unsafe impl Send for FuncRefs {}
+unsafe impl Sync for FuncRefs {}
```

Comment:
Like the above comments, I think this should be removed in favor of something more targeted ideally.